File: HelpfulSense/app/src/main/java/com/andruid/magic/helpfulsense/fragment/AlertFragment.java
package com.andruid.magic.helpfulsense.fragment;

import android.Manifest;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.andruid.magic.helpfulsense.R;
import com.andruid.magic.helpfulsense.activity.IntroActivity;
import com.andruid.magic.helpfulsense.adapter.ActionAdapter;
import com.andruid.magic.helpfulsense.databinding.FragmentAlertBinding;
import com.andruid.magic.helpfulsense.eventbus.ActionEvent;
import com.andruid.magic.helpfulsense.model.Action;
import com.andruid.magic.helpfulsense.model.ActionHolder;
import com.andruid.magic.helpfulsense.util.DialogUtil;
import com.andruid.magic.helpfulsense.util.HolderUtil;
import com.andruid.magic.helpfulsense.util.IntentUtil;
import com.andruid.magic.helpfulsense.viewmodel.ActionViewModel;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.List;
import java.util.Objects;

import eu.davidea.flexibleadapter.helpers.EmptyViewHelper;
import permissions.dispatcher.NeedsPermission;
import permissions.dispatcher.OnNeverAskAgain;
import permissions.dispatcher.OnPermissionDenied;
import permissions.dispatcher.OnShowRationale;
import permissions.dispatcher.PermissionRequest;
import permissions.dispatcher.RuntimePermissions;
import timber.log.Timber;

import static com.andruid.magic.helpfulsense.data.Constants.ACTION_ADD;
import static com.andruid.magic.helpfulsense.data.Constants.ACTION_DIALOG_CANCEL;
import static com.andruid.magic.helpfulsense.data.Constants.ACTION_EDIT;
import static com.andruid.magic.helpfulsense.data.Constants.ACTION_SMS;
import static com.andruid.magic.helpfulsense.data.Constants.ACTION_SWIPE;

@RuntimePermissions
public class AlertFragment extends Fragment implements ActionAdapter.SwipeListener {
    private FragmentAlertBinding binding;
    private ActionAdapter actionAdapter;
    private ActionViewModel actionViewModel;
    private int swipedPos;

    public static AlertFragment newInstance() {
        return new AlertFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        EventBus.getDefault().register(this);
        actionViewModel  = ViewModelProviders.of(this).get(ActionViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_alert, container,
                false);
        binding.emptyLayout.emptyView.setOnClickListener(v ->
                DialogUtil.openAddActionDialog(getChildFragmentManager(),
                        getString(R.string.add_action), ACTION_ADD, null)
        );
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        binding.recyclerView.setItemAnimator(new DefaultItemAnimator());
        setUpViewModel();
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding.unbind();
    }

    private void setUpViewModel() {
        actionViewModel.getSavedActions().observe(this, actions -> {
            actionAdapter = new ActionAdapter(actions, AlertFragment.this);
            binding.recyclerView.setAdapter(actionAdapter);
            actionAdapter.setLongPressDragEnabled(true)
                    .setSwipeEnabled(true);
            EmptyViewHelper.create(actionAdapter, binding.emptyLayout.emptyView);
        });
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onActionEvent(ActionEvent actionEvent){
        Action action = actionEvent.getAction();
        String command = actionEvent.getCommand();
        if(ACTION_ADD.equals(command))
            actionAdapter.addItem(new ActionHolder(action));
        else if(ACTION_EDIT.equals(command))
            actionAdapter.updateItem(swipedPos, new ActionHolder(action), null);
        else if(ACTION_DIALOG_CANCEL.equals(command))
            actionAdapter.notifyDataSetChanged();
        else if(ACTION_SMS.equals(command))
            AlertFragmentPermissionsDispatcher.sendSMSWithPermissionCheck(this, action);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        AlertFragmentPermissionsDispatcher.onRequestPermissionsResult(this, requestCode, grantResults);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_action, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.menu_add_action:
                DialogUtil.openAddActionDialog(getChildFragmentManager(),
                        getString(R.string.add_action), ACTION_ADD, null);
                break;
            case R.id.menu_help:
                startActivity(new Intent(getContext(), IntroActivity.class));
                break;
        }
        return true;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
        if(actionAdapter == null)
            return;
        List<Action> actions = HolderUtil.getActionsFromActionHolders(actionAdapter.getCurrentItems());
        if(!actions.isEmpty())
            actionViewModel.updateSavedActions(actions);
        actionAdapter.release();
    }

    @Override
    public void onSwipe(int position, int direction) {
        Timber.d("swiped: %d", position);
        swipedPos = position;
        if(direction == ItemTouchHelper.LEFT || position == ItemTouchHelper.START) {
            actionAdapter.removeItem(position);
            Timber.d("dir: left");
        } else {
            Timber.d("dir: right");
            DialogUtil.openAddActionDialog(getChildFragmentManager(), getString(R.string.edit_action),
                    ACTION_SWIPE, Objects.requireNonNull(actionAdapter.getItem(position)).getAction());
        }
    }

    @NeedsPermission(Manifest.permission.SEND_SMS)
    void sendSMS(Action action) {
        Intent intent = IntentUtil.buildServiceSmsIntent(getContext(), action.getMessage());
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            Objects.requireNonNull(getContext()).startForegroundService(intent);
        else
            Objects.requireNonNull(getContext()).startService(intent);
    }

    @OnShowRationale(Manifest.permission.SEND_SMS)
    void showRationale(PermissionRequest request){
        DialogUtil.buildInfoDialog(getContext(), getString(R.string.sms_permission), request)
                .show();
    }

    @OnPermissionDenied(Manifest.permission.SEND_SMS)
    void showDenied(){
        Toast.makeText(getContext(), "Denied permission", Toast.LENGTH_SHORT).show();
    }

    @OnNeverAskAgain(Manifest.permission.SEND_SMS)
    void showSettingsDialog(){
        DialogUtil.buildSettingsDialog(getContext(), getString(R.string.sms_permission))
                .show();
    }
}

File: HelpfulSense/app/src/main/java/com/andruid/magic/helpfulsense/ui/fragment/AlertFragment.kt
package com.andruid.magic.helpfulsense.ui.fragment

import android.Manifest
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import com.andruid.magic.eezetensions.startFgOrBgService
import com.andruid.magic.helpfulsense.R
import com.andruid.magic.helpfulsense.data.ACTION_ADD
import com.andruid.magic.helpfulsense.data.ACTION_DIALOG_CANCEL
import com.andruid.magic.helpfulsense.data.ACTION_EDIT
import com.andruid.magic.helpfulsense.data.ACTION_SMS
import com.andruid.magic.helpfulsense.database.DbRepository
import com.andruid.magic.helpfulsense.database.entity.Action
import com.andruid.magic.helpfulsense.database.entity.toPhoneNumbers
import com.andruid.magic.helpfulsense.databinding.FragmentAlertBinding
import com.andruid.magic.helpfulsense.eventbus.ActionEvent
import com.andruid.magic.helpfulsense.ui.activity.HomeActivity
import com.andruid.magic.helpfulsense.ui.activity.IntroActivity
import com.andruid.magic.helpfulsense.ui.adapter.ActionAdapter
import com.andruid.magic.helpfulsense.ui.custom.ItemClickListener
import com.andruid.magic.helpfulsense.ui.dragdrop.DragCallback
import com.andruid.magic.helpfulsense.ui.util.buildInfoDialog
import com.andruid.magic.helpfulsense.ui.util.buildSettingsDialog
import com.andruid.magic.helpfulsense.ui.viewbinding.viewBinding
import com.andruid.magic.helpfulsense.ui.viewmodel.ActionViewModel
import com.andruid.magic.locationsms.util.buildServiceSmsIntent
import com.andruid.magic.locationsms.util.toast
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import permissions.dispatcher.*
import timber.log.Timber

/**
 * Fragment to show all added actions in the database
 */
@RuntimePermissions
class AlertFragment : Fragment(R.layout.fragment_alert), DragCallback.StartDragListener {
    private val binding by viewBinding(FragmentAlertBinding::bind)
    private val touchHelper by lazy {
        val callback = DragCallback(requireContext(), actionAdapter, false)
        ItemTouchHelper(callback)
    }
    private val actionAdapter by lazy {
        ActionAdapter(this, { fromPosition, toPosition ->
            lifecycleScope.launch { updateOrder(fromPosition, toPosition) }
        }) { position, direction ->
            handleSwipe(position, direction)
        }
    }

    private val actionViewModel by viewModels<ActionViewModel>()
    private val actionModeCallback = object : ActionMode.Callback {
        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            actionMode = mode
            mode.menuInflater.inflate(R.menu.menu_action, menu)
            actionAdapter.showDragHandles()
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu) = false

        override fun onDestroyActionMode(mode: ActionMode) {
            actionAdapter.hideDragHandles()
            actionMode = null
        }

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
            hideActionMode()
            return true
        }
    }

    private var actionMode: ActionMode? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        EventBus.getDefault().register(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initRecyclerView()

        actionViewModel.actionLiveData.observe(viewLifecycleOwner) { actions ->
            actionAdapter.submitList(actions)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        onRequestPermissionsResult(requestCode, grantResults)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_action, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_add_action -> findNavController().navigate(AlertFragmentDirections.actionAlertFragmentToMenuAddAction(ACTION_ADD, null))
            R.id.menu_help -> startActivity(Intent(context, IntroActivity::class.java))
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
        Timber.tag("actionLog").d("event bus unregister alert fragment")
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onActionEvent(actionEvent: ActionEvent) {
        Timber.tag("actionLog").d("action event: ${actionEvent.command}")
        actionEvent.action?.let {
            when (actionEvent.command) {
                ACTION_ADD, ACTION_EDIT -> lifecycleScope.launch(Dispatchers.IO) {
                    Timber.tag("actionLog").d("action event: add/edit for ${it.message}")
                    DbRepository.insert(it, actionEvent.command == ACTION_EDIT)
                }
                ACTION_DIALOG_CANCEL -> actionAdapter.notifyDataSetChanged()
                ACTION_SMS -> sendSMSWithPermissionCheck(it)
            }
            null
        }
    }

    /**
     * Send SMS with the selected action message
     * @param action selected action
     */
    @NeedsPermission(Manifest.permission.SEND_SMS, Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.READ_PHONE_STATE)
    fun sendSMS(action: Action) {
        lifecycleScope.launch {
            DbRepository.fetchAllContacts()
                    .flowOn(Dispatchers.Default)
                    .map { contacts -> contacts.toPhoneNumbers() }
                    .collect { phoneNumbers ->
                        val intent = requireContext().buildServiceSmsIntent(action.message, phoneNumbers,
                                HomeActivity::class.java.name, R.mipmap.ic_launcher)
                        startFgOrBgService(intent)
                    }
        }
    }

    @OnShowRationale(Manifest.permission.SEND_SMS, Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.READ_PHONE_STATE)
    fun showRationale(request: PermissionRequest) {
        buildInfoDialog(R.string.sms_permission, request).show()
    }

    @OnPermissionDenied(Manifest.permission.SEND_SMS, Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.READ_PHONE_STATE)
    fun showDenied() {
        toast("Denied permission")
    }

    @OnNeverAskAgain(Manifest.permission.SEND_SMS, Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.READ_PHONE_STATE)
    fun showSettingsDialog() {
        buildSettingsDialog(R.string.sms_permission).show()
    }

    private fun initRecyclerView() {
        binding.recyclerView.apply {
            touchHelper.attachToRecyclerView(this)

            adapter = actionAdapter
            itemAnimator = DefaultItemAnimator()
            setEmptyViewClickListener {
                val navController = findNavController()
                navController.navigate(AlertFragmentDirections.actionAlertFragmentToMenuAddAction(ACTION_ADD, null))
            }
            addOnItemTouchListener(object : ItemClickListener(requireContext(), this) {
                override fun onLongClick(view: View, position: Int) {
                    super.onLongClick(view, position)
                    if (actionMode == null)
                        showActionMode()
                }
            })
        }
    }

    private fun hideActionMode() {
        actionMode?.finish()
    }

    private fun showActionMode() {
        if (actionMode == null)
            (requireActivity() as AppCompatActivity).startSupportActionMode(actionModeCallback)
    }

    private suspend fun updateOrder(fromPos: Int, toPos: Int) {
        val actions = actionAdapter.currentList.toList()
        if (fromPos < toPos) {
            for (i in fromPos..toPos) {
                val action = actions[i]
                DbRepository.insert(action.copy(order = i), true)
            }
        } else if (fromPos > toPos) {
            for (i in fromPos downTo toPos) {
                val action = actions[i]
                DbRepository.insert(action.copy(order = i), true)
            }
        }
    }

    private fun handleSwipe(position: Int, direction: Int) {
        actionAdapter.getItemAtPosition(position)?.let { action ->
            when (direction) {
                ItemTouchHelper.LEFT, ItemTouchHelper.START ->
                    lifecycleScope.launch(Dispatchers.IO) { DbRepository.delete(action) }
                else -> {
                    actionAdapter.notifyItemChanged(position)
                    val navController = findNavController()
                    navController.navigate(AlertFragmentDirections.actionAlertFragmentToMenuAddAction(ACTION_EDIT, action))
                }
            }
        }
    }

    override fun requestDrag(viewHolder: RecyclerView.ViewHolder) {
        touchHelper.startDrag(viewHolder)
    }
}